src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1702,145 +1855,220 @@ final byte getCoder() {
         return COMPACT_STRINGS ? coder : UTF16;
     }
 
+    // Package access for String and StringBuffer.
     final boolean isLatin1() {
+        return isLatin1(coder);
+    }
+
+    private static boolean isLatin1(byte coder) {
         return COMPACT_STRINGS && coder == LATIN1;
     }
 
-    private final void putCharsAt(int index, char[] s, int off, int end) {
-        if (isLatin1()) {
-            byte[] val = this.value;
-            for (int i = off, j = index; i < end; i++) {
+    /**
+     * {@return Return the buffer containing the composed string and inserted characters}
+     * If the buffer coder needs to support UTF16 and does not, it is inflated and a different
+     * buffer is returned. The caller is responsible for setting the coder and updating the value ref
+     * based solely on the difference in the buffer reference.
+     *
+     * @param value byte array destination for the string (if the coder matches)
+     * @param coder the original buffer coder
+     * @param count the count of characters in the original buffer
+     * @param index the insertion point
+     * @param s char[] array to insert from
+     * @param off offset of the first character
+     * @param end the offset of the last character (exclusive)
+     */
+    private static byte[] putCharsAt(byte[] value, byte coder, int count, int index, char[] s, int off, int end) {
+        if (isLatin1(coder)) {
+            int latin1Len = StringUTF16.compress(s, off, value, index, end - off);
+            for (int i = off + latin1Len, j = index + latin1Len; i < end; i++) {
                 char c = s[i];
                 if (StringLatin1.canEncode(c)) {
-                    val[j++] = (byte)c;
+                    value[j++] = (byte)c;
                 } else {
-                    inflate();
-                    StringUTF16.putCharsSB(this.value, j, s, i, end);
-                    return;
+                    value = inflateToUTF16(value, count);
+                    // Store c to make sure sb has a UTF16 char
+                    StringUTF16.putCharSB(value, j++, c);
+                    i++;
+                    StringUTF16.putCharsSB(value, j, s, i, end);
+                    return value;
                 }
             }
         } else {
-            StringUTF16.putCharsSB(this.value, index, s, off, end);
+            StringUTF16.putCharsSB(value, index, s, off, end);
         }
+        return value;
     }
```

Comment:
I think the logic can be simplified, without a need for the loop.
```
    private static byte[] putCharsAt(byte[] value, byte coder, int count, int index, char[] s, int off, int end) {
        if (isLatin1(coder)) {
            int latin1Len = StringUTF16.compress(s, off, value, index, end - off);
            if (latin1Len < end - off) {
                value = inflateToUTF16(value, count);
                StringUTF16.putCharsSB(value, index + latin1Len, s, off + latin1Len, end);
            }
        } else {
            StringUTF16.putCharsSB(value, index, s, off, end);
        }
        return value;
    }
```

Reply:
Unfortunately, my simplification isn't sound when `s` is concurrently modified.
Please disregard.